File: apps/api/src/services/order-service.ts
import { eq, and } from "drizzle-orm";
import type { Database } from "../db/index.js";
import {
  orders,
  orderItems,
  meals,
  type OrderRow,
  type OrderItemRow,
} from "../db/schema/index.js";

export interface CreateOrderData {
  items: Array<{ mealId: string; quantity: number }>;
  deliverySlotId?: string;
  deliveryAddress?: string;
}

function rowToOrder(row: OrderRow, items: OrderItemRow[]) {
  return {
    id: row.id,
    userId: row.userId,
    status: row.status,
    type: row.type,
    total: row.total,
    deliverySlotId: row.deliverySlotId ?? undefined,
    deliveryAddress: row.deliveryAddress ?? undefined,
    posterOrderId: row.posterOrderId ?? undefined,
    stripePaymentIntentId: row.stripePaymentIntentId ?? undefined,
    createdAt: row.createdAt,
    updatedAt: row.updatedAt,
    items: items.map((item) => ({
      id: item.id,
      orderId: item.orderId,
      mealId: item.mealId,
      mealName: item.mealName,
      quantity: item.quantity,
      unitPrice: item.unitPrice,
    })),
  };
}

export async function createOrder(
  db: Database,
  userId: string,
  data: CreateOrderData
) {
  // 1. Validate all mealIds exist and are active
  const mealIds = data.items.map((i) => i.mealId);
  const mealRows = await db
    .select()
    .from(meals)
    .where(eq(meals.active, true));

  const activeMealMap = new Map(mealRows.map((m) => [m.id, m]));

  for (const item of data.items) {
    if (!activeMealMap.has(item.mealId)) {
      throw Object.assign(
        new Error(`Meal not found or inactive: ${item.mealId}`),
        { statusCode: 400, code: "INVALID_MEAL" }
      );
    }
  }

  // 2. Calculate total
  let total = 0;
  for (const item of data.items) {
    const meal = activeMealMap.get(item.mealId)!;
    total += meal.price * item.quantity;
  }

  // 3. Insert order
  const [orderRow] = await db
    .insert(orders)
    .values({
      userId,
      status: "pending_payment",
      type: data.deliverySlotId ? "delivery" : "pickup",
      total,
      deliverySlotId: data.deliverySlotId ?? null,
      deliveryAddress: data.deliveryAddress ?? null,
    })
    .returning();

  if (!orderRow) throw new Error("Failed to create order");

  // 4. Insert order items
  const itemValues = data.items.map((item) => {
    const meal = activeMealMap.get(item.mealId)!;
    return {
      orderId: orderRow.id,
      mealId: item.mealId,
      mealName: meal.name,
      quantity: item.quantity,
      unitPrice: meal.price,
    };
  });

  const insertedItems = await db
    .insert(orderItems)
    .values(itemValues)
    .returning();

  return rowToOrder(orderRow, insertedItems);
}

export async function getOrder(db: Database, orderId: string) {
  const [orderRow] = await db
    .select()
    .from(orders)
    .where(eq(orders.id, orderId))
    .limit(1);

  if (!orderRow) return null;

  const itemRows = await db
    .select()
    .from(orderItems)
    .where(eq(orderItems.orderId, orderId));

  return rowToOrder(orderRow, itemRows);
}

export async function listUserOrders(db: Database, userId: string) {
  const orderRows = await db
    .select()
    .from(orders)
    .where(eq(orders.userId, userId))
    .orderBy(orders.createdAt);

  const results = [];
  for (const orderRow of orderRows) {
    const itemRows = await db
      .select()
      .from(orderItems)
      .where(eq(orderItems.orderId, orderRow.id));
    results.push(rowToOrder(orderRow, itemRows));
  }

  return results;
}

export async function updateOrderStatus(
  db: Database,
  orderId: string,
  status: string
) {
  const [row] = await db
    .update(orders)
    .set({ status, updatedAt: new Date() })
    .where(eq(orders.id, orderId))
    .returning();

  if (!row) return null;

  const itemRows = await db
    .select()
    .from(orderItems)
    .where(eq(orderItems.orderId, orderId));

  return rowToOrder(row, itemRows);
}

export async function setOrderPosterOrderId(
  db: Database,
  orderId: string,
  posterOrderId: string
) {
  const [row] = await db
    .update(orders)
    .set({ posterOrderId, updatedAt: new Date() })
    .where(eq(orders.id, orderId))
    .returning();

  if (!row) return null;

  const itemRows = await db
    .select()
    .from(orderItems)
    .where(eq(orderItems.orderId, orderId));

  return rowToOrder(row, itemRows);
}

export async function setOrderStripePaymentIntentId(
  db: Database,
  orderId: string,
  stripePaymentIntentId: string
) {
  const [row] = await db
    .update(orders)
    .set({ stripePaymentIntentId, updatedAt: new Date() })
    .where(eq(orders.id, orderId))
    .returning();

  if (!row) return null;

  const itemRows = await db
    .select()
    .from(orderItems)
    .where(eq(orderItems.orderId, orderId));

  return rowToOrder(row, itemRows);
}

File: backend/app/services/order_service.py
"""Order service — create, list, get, update status."""

import uuid

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models.meal import Meal
from app.models.order import Order, OrderItem
from app.schemas.order import OrderCreate
from app.services.pricing_service import calculate_item_price


async def create_order(
    db: AsyncSession, user_id: uuid.UUID, data: OrderCreate
) -> Order:
    # Validate meals and snapshot prices
    meal_ids = [item.meal_id for item in data.items]
    result = await db.execute(
        select(Meal).where(Meal.id.in_(meal_ids), Meal.active.is_(True))
    )
    meals_by_id = {m.id: m for m in result.scalars().all()}

    missing = [str(mid) for mid in meal_ids if mid not in meals_by_id]
    if missing:
        raise ValueError(f"Meals not found or inactive: {', '.join(missing)}")

    order_items = []
    total = 0.0
    for item in data.items:
        meal = meals_by_id[item.meal_id]

        # Validate negative extras don't exceed meal's macros
        if meal.protein + item.extra_protein < 0:
            raise ValueError(
                f"Cannot remove more protein than {meal.name} contains"
            )
        if meal.carbs + item.extra_carbs < 0:
            raise ValueError(
                f"Cannot remove more carbs than {meal.name} contains"
            )
        if meal.fat + item.extra_fat < 0:
            raise ValueError(
                f"Cannot remove more fat than {meal.name} contains"
            )

        unit_price = await calculate_item_price(
            db, meal, item.extra_protein, item.extra_carbs, item.extra_fat
        )
        line_total = unit_price * item.quantity
        total += line_total
        order_items.append(
            OrderItem(
                meal_id=meal.id,
                meal_name=meal.name,
                quantity=item.quantity,
                unit_price=unit_price,
                extra_protein=item.extra_protein,
                extra_carbs=item.extra_carbs,
                extra_fat=item.extra_fat,
            )
        )

    order = Order(
        user_id=user_id,
        status="pending_payment",
        type=data.type,
        total=round(total, 2),
        delivery_slot_id=data.delivery_slot_id,
        delivery_address=data.delivery_address,
        items=order_items,
    )
    db.add(order)
    await db.commit()
    await db.refresh(order, attribute_names=["items"])
    return order


async def get_order(
    db: AsyncSession, order_id: uuid.UUID
) -> Order | None:
    result = await db.execute(
        select(Order)
        .options(selectinload(Order.items))
        .where(Order.id == order_id)
    )
    return result.scalar_one_or_none()


async def list_orders(
    db: AsyncSession, user_id: uuid.UUID
) -> list[Order]:
    result = await db.execute(
        select(Order)
        .options(selectinload(Order.items))
        .where(Order.user_id == user_id)
        .order_by(Order.created_at.desc())
    )
    return list(result.scalars().all())


async def update_order_status(
    db: AsyncSession, order_id: uuid.UUID, new_status: str
) -> Order | None:
    order = await get_order(db, order_id)
    if order is None:
        return None
    order.status = new_status
    await db.commit()
    await db.refresh(order)
    return order
